sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -0,0 +1,1396 @@
+semantics and buffer synchronization rules in the SYCL specification. It will be
+extended either for the lifetime of the graph (including both modifiable graphs
+and the executable graphs created from them) or until the buffer is no longer
+required by the graph (such as after being replaced through executable graph update).
+
+Because of the extension of storage lifetimes, users should avoid the use of the
+buffer copy-back on destruction mechanism. If used in code intended to be
+executed as part of a graph, it may not perform as expected.
+
+=== Buffer Limitations for Record & Replay API
+
+Because of the delayed execution of a recorded graph, it is not possible to support
+captured code which relies on the copy-back on destruction behavior of buffers.
+Typically, applications would rely on this behavior to do work on the host which
+cannot inherently be captured inside a command graph. Thus, when recording to a graph
+it is an error to submit a command which has an accessor on a buffer which would
+cause a write-back to happen. Using an incompatible buffer in this case will result
+in a synchronous error being thrown with error code `invalid`.
+
+The copy-back mechanism can be disabled explicitly for buffers with attached host
+storage using either `buffer::set_final_data(nullptr)` or
+`buffer::set_copy_back(false)`.
+
+It is also an error to create a host accessor to a buffer which is used in
+commands which are currently being recorded to a command graph. Attempting to
+construct a host accessor to an incompatible buffer will result in a
+synchronous error being thrown with error code `invalid`.
+
+=== Host Tasks
+
+:host-task: https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#subsec:interfaces.hosttasks
+:cg-scope: https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:command.group.scope
+
+A {host-task}[host task] is a native C++ callable, scheduled according to SYCL
+dependency rules. It is valid to record a host task as part of graph, though it
+may lead to sub-optimal graph performance because a host task node may prevent
+the SYCL runtime from submitting the entire executable `command_graph` to the
+device at once.
+
+Host tasks can be updated as part of <<executable-graph-update, executable graph update>>
+by replacing the whole node with the new callable.
+
+=== Command Group Function Evaluation
+
+Host code within a command group function object is evaluated when the command
+group is added to a graph. This is either before the return of the call to
+`command_graph::add()` when using the explicit API or before the return of the call to
+`queue::submit()` when submitting a command group to a queue that is recording to a graph.
+This behaviour is in keeping with the existing {cg-scope}[command group] behaviour but may have
+implications for command group functions containing arbitrary host code. This could
+affect the behaviour of captured code due to the delayed execution of commands.
```

Comment:
It's not clear what problem you are concerned about here.  Even in the core spec (without graphs), host code in the command group function is executed synchronously and code in the command is delayed (executed asynchronously).  Why are things any different with the graph API?

Reply:
It is perhaps an uncommon usage but consider this example:

```c++
...
auto usm_ptr = sycl::malloc_shared<float>(1, q);
q.submit([&](sycl::handler& h){
    //Modify the contents of usm_ptr here before kernel is executed
    h.single_task([&](){
        // Do something with the modified contents of usm_ptr here
    };
});

// Repeat this for multiple similar submissions
...
```
Without graphs this works as expected, each kernel sees the correctly modified data at the time of execution. With graphs all of these handlers will be evaluated first, each modifying the data, then all the kernels will be executed. But the behaviour of the program will be different in this case.

Reply:
> Without graphs this works as expected, each kernel sees the correctly modified data at the time of execution.

This is not guaranteed by the SYCL spec.  There is no call to "wait" after the first submit, so it is very likely that all the handlers will execute before any of the commands execute.

Reply:
> > Without graphs this works as expected, each kernel sees the correctly modified data at the time of execution.
> 
> This is not guaranteed by the SYCL spec. There is no call to "wait" after the first submit, so it is very likely that all the handlers will execute before any of the commands execute.

Apologies in my example I was assuming that the kernels had dependencies set via `hander::depends_on()` or some other mechanism so they would execute in order. I just omitted it for simplicity.

Reply:
Adding dependencies between the commands still doesn't provide any guaranteed ordering between host code (in the command handler) vs. device code (the commands themselves).  Even if there were dependencies as you propose, it is still very likely that all the handlers will execute before any of the commands execute.

Reply:
tl;dr; consider and chose the timing of when the cgh_lambda should be executed -- either during the call that adds a node to the graph (too early, but same as `queue::submit`) or when that graph node is executed (fits graph semantics, but later than `queue::submit` would do it).

Both options probably need a detailed explanation of what works and what will give surprising results.

----

I had a similar problem/realisation when implementing my multiplex_queue prototype.

I decided to use `std::future<sycl::event>` as the return type from `mxq::enqueue` and I provide `mxq::will_depend_on(sycl::handler, std::future<sycl::event>)` to support a pending call to `h.depends_on(sycl::event)`.

When it comes time for the multiplex_queue object to actually launch this action, it executes the lambda, the `will_depends_on` code does `h.depends_on(future::get)` and everything works out fine.

However, this only works out fine if the future is still in scope when the multiplex_queue chooses to use it -- I create an array of futures for all the deferred actions and keep it around until the multiplex_queue tells me it is done. Indexing into this array uses the loop variable, which must be captured by value (copied) otherwise it changes before the lambda executes and reads it.

The usage model I implemented first was using sycl::accessor objects, which just worked without me even being aware of the complexity of why it worked. Looking back, I can now see all the ducks lined up in a row. The `sycl::buffer` objects are in scope during the entire lifetime of the multiplex_queue (of course they are), the `sycl::accessor` objects are all local stack variables inside each lambda (equivalent of by-value copies), and SYCL creates all the dependences/events it needs based on the construction/destruction ordering of the `sycl::accessor` objects, which is identical to the execution ordering of the cgh_lambda stored by each `mxq::enqueue` call.

The execution pattern for SYCL (queue.submit -> cgh_lambda -> h.parallel_for -> kernel_lambda) is understandable and predictable, but the simple/normal/common usage of `sycl::queue` works for non-trivial reasons, which most customers will never figure out and won't want to care about.

I think a better interface for multiplex_queue would be to insert the linear/stream dependences automatically, and make the user see/deal with this complexity only when they wish to "cross the streams" and add extra event-based dependences.

Reply:
We've removed the text about unexpected behaviour in this PR: https://github.com/reble/llvm/pull/125.